### ISSUE TITLE
util/log: TCP timeout in network logging causes node/cluster unavailability

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#67945]
**Describe the problem**

Found by @logston @chrisseto : if the target address of a network logging configuration  does not have a server responding (i.e. it gets ICMP unreachable / TCP timeout, not TCP RST), that causes log calls to take a very long time (standard TCP timeout, which I think is 30s) and block server processing.

**To Reproduce**

Use a bogus IP address on the local network in a network logging config.

**Expected behavior**

- Connection establishment is asynchronous
- Until connection is established, logging to the network sink fails immediately with a "not available" error

**Additional data / screenshots**

```
fluent:tcp://10.4.0.6:5170: error dialing network logger: dial tcp 10.4.0.6:5170: i/o timeout
{"tag":"cockroach.ops","channel_numeric":1,"channel":"OPS","timestamp":"1626892393.760254850","cluster_id":"963eafd1-0d7a-4552-bf22-44ffdec49f57","tenant_id":"2","instance_id":10001,"severity_numeric":3,"severity"
:"ERROR","goroutine":421,"file":"util/log/event_log.go","line":46,"entry_counter":0,"redactable":1,"message":"logging error: dial tcp ‹10.4.0.6:5170›: ‹i/o timeout›"}
I210721 18:33:18.761070 421 4@util/log/event_log.go:32  [sql,peer=127.0.0.1:57634,hostssl,client=127.0.0.1:57634,user=root] 54 ={"Timestamp":1626892398761066521,"EventType":"client_authentication_ok","InstanceID":
10001,"Network":"tcp","RemoteAddress":"127.0.0.1:57634","Transport":"hostssl","User":"root","Method":"cert-password"}
fluent:tcp://10.4.0.6:5170: error dialing network logger: dial tcp 10.4.0.6:5170: i/o timeout
{"tag":"cockroach.sessions","channel_numeric":4,"channel":"SESSIONS","timestamp":"1626892398.761070264","cluster_id":"963eafd1-0d7a-4552-bf22-44ffdec49f57","tenant_id":"2","instance_id":10001,"severity_numeric":1,
"severity":"INFO","goroutine":421,"file":"util/log/event_log.go","line":32,"entry_counter":7,"redactable":1,"tags":{"sql":"","peer":"‹127.0.0.1:57634›","hostssl":"","client":"‹127.0.0.1:57634›","user":"root"},"eve
nt":{"Timestamp":1626892398761066521,"EventType":"client_authentication_ok","InstanceID":10001,"Network":"tcp","RemoteAddress":"‹127.0.0.1:57634›","Transport":"hostssl","User":"‹root›","Method":"cert-password"}}
E210721 18:33:23.761400 421 1@util/log/event_log.go:46 ⋮ [-]   logging error: dial tcp ‹10.4.0.6:5170›: ‹i/o timeout›
I210721 18:33:26.881678 63 2@server/status/runtime.go:569  [-] 55  runtime stats: 128 MiB RSS, 96 goroutines (stacks: 1.4 MiB), 25 MiB/45 MiB Go alloc/total (heap fragmentation: 5.8 MiB, heap reserved: 5.0 MiB, he
ap released: 27 MiB), 1.0 MiB/2.1 MiB CGO alloc/total (0.1 CGO/sec), 6.0/5.8 %(u/s)time, 0.0 %gc (0x), 35 KiB/88 KiB (r/w)net
fluent:tcp://10.4.0.6:5170: error dialing network logger: dial tcp 10.4.0.6:5170: i/o timeout
{"tag":"cockroach.ops","channel_numeric":1,"channel":"OPS","timestamp":"1626892403.761400059","cluster_id":"963eafd1-0d7a-4552-bf22-44ffdec49f57","tenant_id":"2","instance_id":10001,"severity_numeric":3,"severity"
:"ERROR","goroutine":421,"file":"util/log/event_log.go","line":46,"entry_counter":0,"redactable":1,"message":"logging error: dial tcp ‹10.4.0.6:5170›: ‹i/o timeout›"}
I210721 18:33:28.762929 420 4@util/log/event_log.go:32  [sql,peer=127.0.0.1:57634,hostssl,client=127.0.0.1:57634,user=root] 56 ={"Timestamp":1626892408762924313,"EventType":"client_session_end","InstanceID":10001,
"Network":"tcp","RemoteAddress":"127.0.0.1:57634","Duration":30004706293}
I210721 18:33:32.342652 7 1@util/log/log_flush.go:99  [-] 57  hangup received, flushing logs
fluent:tcp://10.4.0.6:5170: error dialing network logger: dial tcp 10.4.0.6:5170: i/o timeout
```

Jira issue: CRDB-8795

**Comment by logston** [comment]
- This does not affect dedicated clusters or KV nodes in host clusters. Both can execute SQL without a connection to their TCP logging sinks.
- This only affects SQL pods in host clusters.

**Comment by logston** [comment]
Hey @knz, any word on this issue? This problem makes FluentBit a critical dependency for serving SQL from serverless clusters. We would love to remove some risk by lowering the dependency of CRDB on FluentBit (or any logging sink).

**Comment by joshimhoff** [comment]
+1. We should fix this soon.

**Comment by abarganier** [comment]
@joshimhoff @logston I am picking this up - I'll gather some thoughts on a solution and get back to you. 

> - This does not affect dedicated clusters or KV nodes in host clusters. Both can execute SQL without a connection to their TCP logging sinks.
> - This only affects SQL pods in host clusters.

Just curious, do we have any more details on why it only affect SQL pods? What's one doing that the other isn't?

**Comment by joshimhoff** [comment]
> EDIT: Already answered https://github.com/cockroachdb/cockroach/issues/67945#issuecomment-953112062 😅. The hosts don't have network logging enabled.

I don't quite understand the answer! @logston, is the reason this doesn't affect KV pods that **in CC** network logging is not enabled on KV / dedicated pods? What do we use instead? Sorry! I have a poor mental model here; plz help me develop a decent one.

Yay to this issue being picked up.

**Comment by knz** [comment]
Servers other than SQL pods log only to disk.

**Comment by joshimhoff** [comment]
> Servers other than SQL pods log only to disk.

Isn't that simply how we chose to configure CC or is it not possible for non-SQL pods to do network logging?

**Comment by knz** [comment]
Network logging is available everywhere if we choose to use it.

And then, if/when we enabled network logging elsewhere than SQL pods, we'll run into the same issue.

In other words: @abarganier the problem exists everywhere and it's a defect inside the crdb code. The thing needs to work when the network collector is temporarily unavailable and doesn't respond.

**Comment by abarganier** [comment]
Makes sense, thanks for that context @knz 🎉

**Comment by joshimhoff** [comment]
Thanks!

**Comment by logston** [comment]
> Just curious, do we have any more details on why it only affect SQL pods? What's one doing that the other isn't?
> EDIT: Already answered https://github.com/cockroachdb/cockroach/issues/67945#issuecomment-953112062 😅. The hosts don't have network logging enabled.

and

> In other words: @abarganier the problem exists everywhere and it's a defect inside the crdb code. The thing needs to work when the network collector is temporarily unavailable and doesn't respon

Hmm, I think there is more to this than that. I ran [a suite of tests](https://cockroachlabs.atlassian.net/browse/CC-5355
) to see if this problem was isolated to just SQL pods and I found that, indeed, the problem only showed up when a SQL pod attempted to connect to FluentBit over TCP. Other pods (dedicated and KV) did not seem to hit this issue even when multiple network logging sinks were configured. 

> I don't quite understand the answer! @logston, is the reason this doesn't affect KV pods that in CC network logging is not enabled on KV / dedicated pods? What do we use instead? Sorry! I have a poor mental model here; plz help me develop a decent one.

Neither do I! See my comment above. When this issue first appeared, we did use a FluentBit sidecar that tailed the logs of interest (auth logs) of the CRDB process (via the shared filesystem between the containers). Since then, those sidecars have been removed and replaced by TCP logging and new host and dedicated clusters are created with TCP logging enabled (rather than the FluentBit sidecar). The tests I pointed to from above used clusters with TCP logging enabled. To be complete though, whether the cluster was using a FluentBit sidecar or TCP logging to forward logs on to FluentBit (and then onto Splunk, S3/GCS), the clusters are also [configured to write logs to disk. ](https://github.com/cockroachlabs/managed-service/blame/6582e9c6156e2f3cf6f1b93e2f96d76c0c5f81a6/operators/cockroach-operator/manifests/internal/yaml/logging/configmap.yaml#L12-L21)

**Comment by knz** [comment]
Let me explain the difference. It's really a variant of a similar issue in crdb's RPC services (#53410).

When the IP address configured as collector has an OS attached to it but no process listener, the TCP connect attempt receives a *fast* error from the remote machine ("connection refused", ie. TCP RST) and the logging code inside crdb fails fast.

When the IP address configured as collector has nothing attached to it; or a firewall rule drops packets sent to it, the TCP connect attempt receives a *slow* error  ("connect timeout", i.e. ICMP unreacheable) and the logging code inside crdb fails slow, and causes disruption.

We're looking at the 2nd case here. This is a problem that needs to be solved anyways: these TCP-fail-slow problem exist in a various number of network situations:
- misconfiguration
- log collector pod gets restarted by k8s to upgrade the fluent software
- (e.g. in self-hosted) network collector is outside of k8s cluster and the network gets partitioned
- a network cable is faulty

etc

Generally we must ensure that this software is robust to _all_ kinds of network failures, not just those where there's a convenient TCP RST sent from the other side.

**Comment by logston** [comment]
Gotcha. Thanks for the detail and the linked ticket @knz! While I understand that the issue is due to an inability to fail fast by the client (CRDB in this case), it's not clear to me why this issue would cause an inability to serve SQL queries. But I guess that's what @abarganier will dig into! Furthermore , its even more odd to me that an inability to serve SQL requests only happens when SQL pods fail to fail fast and not when dedicated or KV pods fail to fail fast. That was the behavior I observed when I ran my tests mentioned above. I'm very curious to see what @abarganier finds!

**Comment by knz** [comment]
> It's not clear to me why this issue would cause an inability to serve SQL queries.

Logging  in crdb is synchronous (currently), so if a logging operation waits 30 seconds, whichever other component that was using logging needs to wait 30 seconds too. Generally, delays in logging is _extremely_ disruptive to crdb.

That's the point of implementing proper buffering. We've made some improvements to buffering in 22.1 in other areas (notably in the HTTP logging sink), however 1) it's not available to Fluent sinks yet 2) the buffering implementation still has shortcomings (listed in 6 other github issues).

**Comment by knz** [comment]
> not when dedicated or KV pods fail to fail fast.

I can readily cause KV failures in crdb if I block logging for more than a few seconds. I'm not sure what test you've performed? This is a general shortcoming, not specific to fluent sinks.

**Comment by logston** [comment]
> I can readily cause KV failures in crdb if I block logging for more than a few seconds.

Oh! Hmm, I wondering what I did different or missed. My tests are recorded [here](https://cockroachlabs.atlassian.net/browse/CC-5355). But if you can recreate the issue in KV pods then clearly there's something more that I need to dig into.

**Comment by knz** [comment]
> What are your thoughts on leveraging this bufferSink for network logging?

For fluent logging you mean? Yeah 💯 that is great idea. However see caveat below.

> It appears this is already supported for both httpSink and fluentSink

That is surprising to me, I didn't think we had plugged this in already. But this great news!

### Caveat

The buffering sink implementation is currently incomplete. That's what prevents us from enabling it by default. There is a list of issues that need to be addressed, listed in [this comment](https://github.com/cockroachdb/cockroach/pull/70330#issuecomment-961421764). That's what's blocking us from considering bufferSink as "complete" (issue #63979).

In particular, issues #72454 , #72459 and #72455 need to be fixed otherwise we'd run into serious production problems.

**Comment by abarganier** [comment]
Thanks for the feedback, glad to hear this is a viable option! 

It sounds like the path to get async network logging working is to start working through these issues (https://github.com/cockroachdb/cockroach/issues/72454 , https://github.com/cockroachdb/cockroach/issues/72459 and https://github.com/cockroachdb/cockroach/issues/72455). This could be a good opportunity for obs-inf to continue familiarizing itself with the logging infra, and its benefits would extend beyond just network logging. 

Let me know if you agree @knz and I can get things moving on our end 👍

**Comment by knz** [comment]
sound like a brilliant plan :+1:

**Comment by logston** [comment]
Hey @abarganier, friendly check in on this one? Is your PR above the only fix required for this ticket? Also, will it land in 21.2's next minor version or in 22.1?

**Comment by abarganier** [comment]
Hey @logston, still working on this. There's a backlog of bugfixes/improvements we need to make to `bufferSink` before we can safely wrap an `httpSink` with one. I'm currently working on https://github.com/cockroachdb/cockroach/issues/72459, after which I need to take care of https://github.com/cockroachdb/cockroach/issues/72455.

@knz I noticed [this comment](https://github.com/cockroachdb/cockroach/pull/70330#issuecomment-961421764) notes issues beyond the three that were [mentioned above](https://github.com/cockroachdb/cockroach/issues/67945#issuecomment-1069403773). Do we need to address these as well before we can get `bufferSink` working with an `httpSink`? Or should we be able to proceed with that once the 3 issues mentioned have been resolved?

**Comment by knz** [comment]
> I'm currently working on https://github.com/cockroachdb/cockroach/issues/72459, after which I need to take care of https://github.com/cockroachdb/cockroach/issues/72455.

The 2nd one is already closed. Did you mean another issue?

> Do we need to address these as well before we can get bufferSink working with an httpSink?

Not *before*, no. That's why I did not choose to mention them above. However we need to get them resolved eventually to call the buffering functionality customer-ready.

**Comment by abarganier** [comment]
> The 2nd one is already closed. Did you mean another issue?

Ah yes, apologies - I meant https://github.com/cockroachdb/cockroach/issues/72459

> However we need to get them resolved eventually to call the buffering functionality customer-ready.

Thanks for the context, understood.

**Comment by mari-crl** [comment]
:information_source: Hello! I am a human and not at all a robot! Look at my very human username! :robot: :notes:
:thinking: Although I tried very hard to figure out what to do with this issue, more powerful human brains will need to help me. (specifically: Both Github and Jira issues were modified after exalate problems) :confounded:
:arrows_counterclockwise: **Please visit this issue's mirror at [CRDB-8795](https://cockroachlabs.atlassian.net/browse/CRDB-8795) and try to sync the two sides up manually.** :star2:
:white_check_mark: When you're finished, comment saying as much asn a member of Developer Infrastructure will be along to finish linking. :link:
:no_entry_sign: Note that until this is done, **this issue is not and will not be synced to Jira with Exalate**. :no_entry_sign:
:sweat_smile: Feeling lost? Don't worry about it! A member of @cockroachdb/exalate-22-cleanup-team will be along shortly to help! :+1:
:construction_worker: Developer Infrastructure members: when ready, open Exalate from the right-hand menu of the mirror issue in Jira, then choose Connect and enter this issue's URN: `cockroachdb/cockroach-67945`. Either way, delete this comment when you're done. :key:
:pray: Thank you for your compliance, my fellow humans! :robot: :wave:

**Comment by abarganier** [comment]
Hey all,

Thanks for your patience. Support for buffered http and fluent sinks is now available on master, and has been backported to v22.1 (https://github.com/cockroachdb/cockroach/pull/83204, https://github.com/cockroachdb/cockroach/issues/72459). Server shutdowns will also now do their best to gracefully flush any buffered log messages to underlying log sinks prior to exiting the process. 

Buffering will also soon be enabled [by default](https://github.com/cockroachdb/cockroach/pull/82893) for http and fluent log sinks on master. We've chosen to not backport these new defaults to avoid changing the logging behavior for 22.1 users using the same config from a previous minor release version. 

The new default buffering configuration for http and fluent sinks on master will be as follows:
```
buffering:
    // The maximum amount of time between flushes to the underlying
    // http or fluent sink.
    max-staleness: 5s
    // `flush-trigger-size` is the size in bytes of accumulated messages
    //  in the buffer which will trigger a flush. 0 disables this trigger.
    flush-trigger-size: 1MiB
    // `max-buffer-size` limits the size of the buffer. When a new message
    // is causing the buffer to overflow beyond this limit, old messages
    // are dropped.
    max-buffer-size: 50MiB
```

cc @andf-crl for docs coverage - we should update our [Configure Logs](https://www.cockroachlabs.com/docs/stable/configure-logs.html) docs to include these config options

**Comment by abarganier** [comment]
Marking closed as the defaults have been updated - @andf-crl let me know if I you need any clarification re: the docs updates.

**Comment by abarganier** [comment]
Just one (good news!) final follow-up - I wanted to see this in action with a serverless cluster, so I spent the day setting up a test environment for that. I verified that enabling buffering for the fluent-servers that tenant pods use fixed the primary concern, which is that a loss in availability of fluent-bit containers would cause tenant SQL servers to become unavailable.

I verified that the SQL servers indeed became unavailable by killing the fluent-bit DaemonSet and explicitly disabling buffered logging on all fluent sinks. It only took a few SQL statements before things started to hang. I then enabled buffered logging, without re-enabling the fluent-bit DaemonSet, restarted the SQL pod, and re-connected. I was then unable to take the SQL server down with normal activity, meaning that the buffering worked as intended.
